### ISSUE TITLE
Problem when an attribute storing object is mutable.

### ISSUE THREAD

**Issue by fingolfin** [gap-system/gap#1371]
While looking at the MatrixObj code, we realized that it is in principle possible that an attribute storing object is mutable.

So if you had a matrix object which is mutable, and also attribute storing, and if `RankMat` is an attribute, then we would store the rank; yet if you later modify the matrix (thus possibly changing the rank), this would not be reflected by the attribute.

It is not quit clear to us how "likely" or easy it is to get a mutable attribute storing object, short of using `SetFilterObj(obj, IsMutable)`; still, there is some concern here about accidental bugs.

So perhaps the default setter code for `IsAttributeStoringRep` could at least check whether the `IsMutable` filter is set, and if so, either print a warning, or error out.

This is not currently done, as visible in this example:
```
gap> G:=Group(  (1,2,3), (1,2) );;
gap> HasSize(G);
false
gap> SetFilterObj(G, IsMutable);
gap> IsMutable(G);
true
gap> Size(G);
6
gap> HasSize(G);
true
```

Note that we cannot just change `IsAttributeStoringRep` to require immutability, because there is no filter `IsImmutable`, just a filter `IsMutable`.

**Comment by stevelinton** [comment]
It is possible in principle that one might have a Mutable object that could still safely store some attributes, because none of the operations that mutate it can change the values of those attributes, but this should be done with explicit setter methods for those attributes.  The generic behaviour introduced by IsAttributeStoringRep is, I think, always the wrong thing for mutable objects. 

I think the best solution would be for the default system setter to simply do nothing when the object is mutable. That way you could create an object with IsMutable and IsAttributeStoringRep  and if it were later made Immutable, it would immediately start storing attributes.

**Comment by stevelinton** [comment]
> I think the best solution would be for the default system setter to simply do nothing when the object is mutable. That way you could create an object with IsMutable and IsAttributeStoringRep and if it were later made Immutable, it would immediately start storing attributes.

I tried this, and it causes numerous test failures. I dug into the first one, and the problem essentially is that when creating a mutable Presentation (for Tietze) the code calls SetOne. (tietze.gi line 381). Since the object has IsPresentationDefaultRep which implies IsAttributeStoringRep the value is stored and later relied upon.

The programmer knows that this is safe, because although the presentation may mutate, the stored attribute value will remain correct. 

On the other hand, I am not sure this is really safe code. It would be perfectly reasonable to define for  a package author to define another attribute of presentations which was not invariant under the allowed mutations. That would also be stored and might be reused when it was no longer valid. 

I think this does need to be sorted out, but it is going to mean mending potentially quite a lot of code which is probably correct in itself, but opens up the possibility of broken extensions.

Maybe we could provide a short way of installing standard attribute setters without the mutability test for individual attributes and object types. So the programmer would still have to take responsibility for confirming that the attribute was safe to set for mutable objects, but make it as easy as possible to do so.

**Comment by stevelinton** [comment]
Returning to this, prompted by #3560, I discover that the problem is not as bad as I thought. The mutable presentation case mentioned above is more or less the only case that shows up in testinstall and is easily worked around, by storing the data in components of the presentation object and then defining methods for `One` and `PrimaryGeneratorsWord` which access it. I'm running more extensive tests now.

**Comment by stevelinton** [comment]
As @ThomasBreuer has pointed out in comments on #3588, there are two (as far as I can see) ways to deal with this, and they have slightly different consequences:

1. Stop the system setters (the default setters for objects in `IsAttributeStoringRep`) from storing anything in mutable objects.

2. Stop all attributes from calling their setters on mutable objects (by modifying `DoAttribute`)

Option 1 allows the programmer to install a custom setter  (when they know that none of the mutators for that object change the value of that particular attribute) which will then operate as normal.

Option 2 allows the programmer to explicitly call the setter (even when it is the default system one) when they want to store an attribute value. 

I can't see a (sensible) way to allow both of these, so, unless someone has a better idea, we have to make a decision.

**Comment by stevelinton** [comment]
Closed by #3592
